### PR TITLE
operands init

### LINKED ISSUE #191: interface BinaryInstruction
In some optimization, like value numbering, can find out BinaryInstruction automatically will be a nice feature. For example:

```go
switch inst.(type) {
case ir.BinaryInstruction:
    opCode := inst.OpCode()
    v1 := getValueOf(inst.Operand(1))
    v2 := getValueOf(inst.Operand(2))
    hashKey := makeHashKey(v1, opCode, v2)
    // ...
}
```

### PR DESCRIPTION
**Pull request by dannypsnl**
resolve #191

### CODE REVIEW

File: ir/inst_other.go [hunk 1]
```diff
@@ -164,6 +172,10 @@ type InstPhi struct {
 	Metadata
 }
 
+func (inst *InstPhi) Operands() []value.Value {
+	panic("implement me")
```
Comment (dannypsnl) [inline ir/inst_other.go:176]: @mewmew some of them, I have no idea what to do, left `panic`

Comment (mewmew) [reply ir/inst_other.go:176]: If there are no value operands, then return nil.

This is what I did for `phi` instructions, not sure if it makes sense or is the best option.

```go
case *ir.InstPhi:
	operands := make([]value.Value, 0, 2*len(inst.Incs))
	for _, inc := range inst.Incs {
		operands = append(operands, inc.X, inc.Pred)
	}
	return operands
```

See https://github.com/llir/llvm/issues/191#issuecomment-881198489 for comparison.

**Comment by dannypsnl** [comment]
@mewmew updated

File: ir/instruction.go [hunk 2]
```diff
@@ -106,4 +108,6 @@ type Instruction interface {
 	// isInstruction ensures that only instructions can be assigned to the
 	// instruction.Instruction interface.
 	isInstruction()
+	// Operands returns a mutable list of operands of the instruction.
```
Comment (mewmew) [inline ir/instruction.go:111]: Should we return a *mutable* list? If so, the return type should probably be `[]*value.Value`?

File: ir/ir_test.go [hunk 3]
```diff
@@ -100,6 +100,8 @@ var (
 	_ Instruction = (*InstBitCast)(nil)
 	_ Instruction = (*InstAddrSpaceCast)(nil)
 	// Other instructions.
+	_ Instruction = (*InstFreeze)(nil)
```
Comment (mewmew) [inline ir/ir_test.go:103]: Good that you added type assertion tests for these instructions!

File: ir/terminator.go [hunk 4]
```diff
@@ -862,6 +908,10 @@ type TermUnreachable struct {
 	Metadata
 }
 
+func (term *TermUnreachable) Operands() []value.Value {
+	return []value.Value{}
```
Comment (mewmew) [inline ir/terminator.go:912]: Use `return nil` for consistency with `InstLandingPad` and `TermSwitch`.

File: ir/inst_memory.go [hunk 5]
```diff
@@ -262,6 +274,10 @@ type InstFence struct {
 	Metadata
 }
 
+func (inst *InstFence) Operands() []value.Value {
+	return []value.Value{}
```
Comment (mewmew) [inline ir/inst_memory.go:278]: Use `return nil` for consistency with `InstLandingPad` and `TermSwitch`.

File: ir/inst_aggregate.go [hunk 6]
```diff
@@ -90,6 +92,8 @@ type InstInsertValue struct {
 	Metadata
 }
 
+func (inst *InstInsertValue) Operands() []value.Value { return []value.Value{inst.X, inst.Elem} }
```
Comment (mewmew) [inline ir/inst_aggregate.go:95]: Use `go fmt` to format the source file.

**Review by mewmew — COMMENTED** [review]
Good job on this. We should decide on the API of `Operators`. Should it return a *mutable* list of operands or not? If it should return a mutable list, then the return type should be updated to `[]*value.Value`.

File: ir/instruction.go (hunk 2 above)
Comment (dannypsnl) [reply ir/instruction.go:111]: Why? Shouldn't interface already a pointer?

Comment (mewmew) [reply ir/instruction.go:111]: The underlying value of a concrete type implementing the `value.Value` interface is often a pointer, e.g. `*ir.InstAdd`. Also, interface types are often treated as pointers, e.g. compare against `nil`.

However, lets consider the use case where you want to *update* the operands of an instruction to other values, then, the following code would not work if using `[]value.Value`.

See example at https://play.golang.org/p/Vy2BnlC7FkD (using `[]value.Value`) and https://play.golang.org/p/ICR50iHBfs9 (using `[]*value.Value`).

Using `updateOperands` (which uses `[]value.Value`) is not able to update the operands (thus not mutable). Note that the `add` instruction was not updated to use `zeroinitializer`:

Output:
```llvm
define i32 @foo(i32 %x, i32 %y) {
entry:
	%0 = add i32 %x, %y
	ret i32 %0
}

define i32 @main() {
entry:
	%0 = call i32 @foo(i32 30, i32 12)
	ret i32 %0
}
```

Using `updateMutableOperands` (which uses `[]*value.Value`) is able to update the operands (thus mutable). Note that the `add` instruction was updated to use `zeroinitializer`:

Output:
```llvm
define i32 @foo(i32 %x, i32 %y) {
entry:
	%0 = add i32 zeroinitializer, zeroinitializer
	ret i32 %0
}

define i32 @main() {
entry:
	%0 = call i32 @foo(i32 30, i32 12)
	ret i32 %0
}
```

File: ir/inst_aggregate.go (hunk 6 above)
Comment (dannypsnl) [reply ir/inst_aggregate.go:95]: I think this is valid under `go fmt`

Comment (dannypsnl) [reply ir/inst_aggregate.go:95]: Because after I run `go fmt`, it still keep this result

File: ir/instruction.go (hunk 2 above)
Comment (dannypsnl) [reply ir/instruction.go:111]: oh, I get what you mean. However, we can update the operand directly via the instruction itself. Thus, I can't see the reason that makes `Operand` more complicated(be mutable). So I would update the document string.

**Comment by mewmew** [comment]
Hi @dannypsnl,

As this will be part of the core API which defines how instructions are used, I think it is important that we take time to consider use cases and get the design just right for the `Operands` method.

I am leaning towards having `Operands` return a mutable list, as it allows for a very powerful API; see https://github.com/llir/llvm/issues/191#issuecomment-883995253 for further background. This would also help resolve (at least part of) the use cases detailed in #42.

> Regarding the first point, I think this will allow for a very powerful API. Similar to the IR traversal API (inspired by Go fix) that we have in irutil.Walk.

So. I would like for us to take time and consider the benefit vs. drawbacks of having a mutable list returned by the `Operands` type, as this may also be used for other data analysis APIs in the future, e.g. use-def chains, etc.

For the time being, those who need to start using the `Operands` API today before we finalize it may do so by pinning the Git commit hash of the *operands* branch (e.g. by doing `go get -u github.com/llir/llvm@operands` in a Go module enabled repository).

Cheers,
Robin

**Comment by dannypsnl** [comment]
Since `value.Value` is an interface, does mutation really work on it?

**Comment by mewmew** [comment]
> Since value.Value is an interface, does mutation really work on it?

No, this is why we would need to use `[]*value.Value` for mutation.

Note, the AST walker of the `go fix` tool relies on pointer to interface to mutate Go source code by AST traversal. It provides a really powerful API!

From [golang/go/src/cmd/fix/walk.go](https://github.com/golang/go/blob/9e85dc5f183d688f5297203dd76f281a6d87d94f/src/cmd/fix/fix.go#L64):

```go
	// pointers to interfaces
	case *ast.Decl:
		walkBeforeAfter(*n, before, after)
	case *ast.Expr:
		walkBeforeAfter(*n, before, after)
	case *ast.Spec:
		walkBeforeAfter(*n, before, after)
	case *ast.Stmt:
		walkBeforeAfter(*n, before, after)

	// pointers to struct pointers
	case **ast.BlockStmt:
		walkBeforeAfter(*n, before, after)
	case **ast.CallExpr:
		walkBeforeAfter(*n, before, after)

	// etc...
```

**Pull request closed without merging**